File: backend/app/schemas/district.py
"""商圈分析 Pydantic Schema。"""
from __future__ import annotations

import uuid
from datetime import datetime
from typing import Literal

from pydantic import BaseModel, ConfigDict, Field


class AnalyzeResponse(BaseModel):
    snapshot_id: uuid.UUID
    poi_total: int
    competitor_count: int
    density_per_km2: float | None
    mapping_status: str
    excluded_self_count: int


class CategoryStat(BaseModel):
    category: str
    count: int


class PoiOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    poi_id: str
    name: str
    category: str | None = None
    typecode: str | None = None
    address: str | None = None
    tel: str | None = None
    tag: str | None = None
    business_area: str | None = None
    rating: float | None = None
    cost: float | None = None
    business_hours: str | None = None
    lng: float | None = None
    lat: float | None = None
    distance_m: int
    is_competitor: bool
    is_competitor_auto: bool = False
    is_competitor_manual: bool = False
    excluded_as_self: bool


class SnapshotSummaryResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    shop_id: uuid.UUID
    center_lng: float | None = None
    center_lat: float | None = None
    geocode_level: str | None = None
    radius_m: int
    poi_total: int
    competitor_count: int
    category_stats: list[CategoryStat] | None = None
    density_per_km2: float | None = None
    mapping_status: str
    status: str
    error_message: str | None = None
    excluded_self_count: int = 0
    created_at: datetime


class SnapshotListResponse(BaseModel):
    items: list[SnapshotSummaryResponse]
    total: int
    page: int
    size: int


class SnapshotDetailResponse(SnapshotSummaryResponse):
    pois: list[PoiOut]


class PoisListResponse(BaseModel):
    items: list[PoiOut]
    total: int
    page: int
    size: int


class CompetitorOut(BaseModel):
    poi_id: str
    name: str
    category: str | None = None
    typecode: str | None = None
    address: str | None = None
    tel: str | None = None
    tag: str | None = None
    business_area: str | None = None
    rating: float | None = None
    cost: float | None = None
    business_hours: str | None = None
    distance_m: int
    lng: float | None = None
    lat: float | None = None
    is_competitor_manual: bool = False


class PoiOverrideUpsert(BaseModel):
    is_competitor: bool
    note: str | None = None
    poi_name: str | None = None


class PoiOverrideOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    poi_id: str
    poi_name: str | None = None
    is_competitor: bool
    note: str | None = None
    updated_at: datetime


class PoiOverrideListResponse(BaseModel):
    items: list[PoiOverrideOut]
    total: int


class MapConfigResponse(BaseModel):
    amap_js_key: str
    proxy_path: str = "/api/v1/district/_AMapService"

File: frontend/src/services/district.ts
// 商圈分析 API 服务 — 对应 docs/contracts/district-api.md v1.0
import api from "./api";

export interface CategoryStat {
  category: string;
  count: number;
}

export type MappingStatus = "full" | "none";
export type SnapshotStatus = "analyzed" | "failed";

export interface SnapshotSummary {
  id: string;
  shop_id: string;
  center_lng: number | null;
  center_lat: number | null;
  geocode_level: string | null;
  radius_m: number;
  poi_total: number;
  competitor_count: number;
  category_stats: CategoryStat[] | null;
  density_per_km2: number | null;
  mapping_status: MappingStatus;
  status: SnapshotStatus;
  error_message: string | null;
  excluded_self_count: number;
  created_at: string;
}

export interface SnapshotListResponse {
  items: SnapshotSummary[];
  total: number;
  page: number;
  size: number;
}

export interface Poi {
  id: string;
  poi_id: string;
  name: string;
  category: string | null;
  typecode: string | null;
  address: string | null;
  tel: string | null;
  tag: string | null;
  business_area: string | null;
  rating: number | null;
  cost: number | null;
  business_hours: string | null;
  lng: number | null;
  lat: number | null;
  distance_m: number;
  is_competitor: boolean;
  is_competitor_auto?: boolean;
  is_competitor_manual?: boolean;
  excluded_as_self: boolean;
}

export interface PoisListResponse {
  items: Poi[];
  total: number;
  page: number;
  size: number;
}

export interface Competitor {
  poi_id: string;
  name: string;
  category: string | null;
  typecode: string | null;
  address: string | null;
  tel: string | null;
  tag: string | null;
  business_area: string | null;
  rating: number | null;
  cost: number | null;
  business_hours: string | null;
  distance_m: number;
  lng: number | null;
  lat: number | null;
  is_competitor_manual?: boolean;
}

export interface PoiOverrideUpsert {
  is_competitor: boolean;
  note?: string | null;
  poi_name?: string | null;
}

export interface PoiOverride {
  poi_id: string;
  poi_name: string | null;
  is_competitor: boolean;
  note: string | null;
  updated_at: string;
}

export interface PoiOverrideListResponse {
  items: PoiOverride[];
  total: number;
}

export interface AnalyzeResponse {
  snapshot_id: string;
  poi_total: number;
  competitor_count: number;
  density_per_km2: number | null;
  mapping_status: MappingStatus;
  excluded_self_count: number;
}

export interface MapConfig {
  amap_js_key: string;
  proxy_path: string;
}

export interface SnapshotDetail extends SnapshotSummary {
  pois: Poi[];
}

export interface SnapshotListParams {
  page?: number;
  size?: number;
  status?: SnapshotStatus;
}

export interface PoiListParams {
  page?: number;
  size?: number;
  include_excluded?: boolean;
}

export const districtService = {
  analyze: (shopId: string) =>
    api.post<AnalyzeResponse>(`/shops/${shopId}/district/analyze`),

  latest: (shopId: string) =>
    api.get<SnapshotSummary>(`/shops/${shopId}/district/latest`),

  listSnapshots: (shopId: string, params?: SnapshotListParams) =>
    api.get<SnapshotListResponse>(`/shops/${shopId}/district/snapshots`, { params }),

  snapshotDetail: (shopId: string, snapshotId: string) =>
    api.get<SnapshotDetail>(`/shops/${shopId}/district/snapshots/${snapshotId}`),

  listPois: (shopId: string, snapshotId: string, params?: PoiListParams) =>
    api.get<PoisListResponse>(`/shops/${shopId}/district/snapshots/${snapshotId}/pois`, {
      params,
    }),

  competitors: (shopId: string, snapshotId: string) =>
    api.get<Competitor[]>(`/shops/${shopId}/district/snapshots/${snapshotId}/competitors`),

  listOverrides: (shopId: string) =>
    api.get<PoiOverrideListResponse>(`/shops/${shopId}/district/poi-overrides`),

  setPoiOverride: (shopId: string, poiId: string, body: PoiOverrideUpsert) =>
    api.put<PoiOverride>(`/shops/${shopId}/district/poi-overrides/${poiId}`, body),

  deletePoiOverride: (shopId: string, poiId: string) =>
    api.delete(`/shops/${shopId}/district/poi-overrides/${poiId}`),

  mapConfig: () => api.get<MapConfig>("/district/map-config"),
};
